### PR TITLE
 Use function pointers instead of marshalled delegates in EventSource

### PR DESCRIPTION
**Pull request by jkotas**
Function pointers are more efficient and avoid JITing. This is one of the last remaining uses of marshalled delegates in core framework. It allows the infrastructure for marshalled delegates to be trimmed.

### CODE REVIEW

File: src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs [hunk 1]
```diff
@@ -443,15 +443,15 @@ private void EmitNativeToInterp(StreamWriter w, ref List<PInvokeCallback> callba
             pindex = 0;
             if (!is_void)
             {
-                sb.Append("&res");
+                sb.Append("(int*)&res");
```
Comment (jkotas) [inline src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:446]: This change introduced `UnmanagedCallersOnly` callback with long argument. It caused wasm build to fail with `pinvoke-table.h(303,87): error G55D6342F: (NETCORE_ENGINEERING_TELEMETRY=Build) incompatible pointer types passing 'int64_t *' (aka 'long long *') to parameter of type 'int *'`. @radical Is this the right fix for this build error?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventProvider.cs [hunk 2]
```diff
@@ -1,47 +1,58 @@
...
+
+            Debug.Assert(!_gcHandle.IsAllocated);
+            _gcHandle = GCHandle.Alloc(this);
+
+            _provHandle = EventPipeInternal.CreateProvider(eventSource.Name, &Callback, (void*)GCHandle.ToIntPtr(_gcHandle));
+            if (_provHandle == 0)
             {
                 // Unable to create the provider.
-                returnStatus = 1;
+                throw new OutOfMemoryException();
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventProvider.cs:36]: Do we need to _gcHandle.Dispose in this case, or is EventUnregister going to be called to handle that?

Comment (noahfalk) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventProvider.cs:36]: EventUnregister is eventually going to be called via EventProvider.Dispose so we should be OK.

**Review by noahfalk — APPROVED** [review]
LGTM

cc @davmason

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventProvider.cs (hunk 2 above)
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventProvider.cs:36]:
```suggestion
                _gcHandle.Free();
                throw new OutOfMemoryException();
```

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventProvider.cs:37]: Still, it does not hurt to free the handle here.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs [hunk 3]
```diff
@@ -1179,79 +1116,74 @@ protected internal unsafe bool WriteEvent(ref EventDescriptor eventDescriptor, I
...
-                callbackContext,
-                ref registrationHandle);
+            uint status = Interop.Advapi32.EventRegister(
+                &providerId,
+                &Callback,
+                (void*)GCHandle.ToIntPtr(_gcHandle),
+                &registrationHandle);
+            if (status != 0)
+            {
+                throw new ArgumentException(Interop.Kernel32.GetMessage((int)status));
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs:1165]:
```suggestion
                _gcHandle.Free();
                throw new ArgumentException(Interop.Kernel32.GetMessage((int)status));
```

File: src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs (hunk 1 above)
Comment (jkotas) [inline src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:446]: @BrzVlad Could you please review the change in this file?

Comment (vargaz) [reply src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:446]: It looks fine.

Comment (lambdageek) [reply src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:446]: @maraf could you take a look

Comment (lambdageek) [reply src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:446]: LGTM, fwiw.  We generate `WasmInterpEntrySig_nnn` to have a signature like `void (*)(int* resultDest, int* arg0, ... int* argN, int* extraArg)`.  So casting all the addresses to `int*` is ok.

**Review by davmason — APPROVED** [review]
LGTM

**Review by maraf — APPROVED** [review]
WASM bits LGTM

**Comment by janvorli** [comment]
@jkotas this PR has broken unloadability for a couple of coreclr tests on both Windows and Unix. The unloading is blocked due to root like this (the root path differs a bit test by test):
```
(lldb) gcroot -all 00007fbf6a813690
HandleTable:
    00007FFFF4AC1390 (strong handle)
    -> 00007FBF6A81A800 System.Diagnostics.Tracing.EventPipeEventProvider
    -> 00007FBF6A81A828 System.Diagnostics.Tracing.EventEnableCallback
    -> 00007FBF6A81A7A8 System.Diagnostics.Tracing.EventSource+OverrideEventProvider
    -> 00007FBF6A816538 System.Threading.Tasks.ParallelEtwProvider
    -> 00007FBF6A813690 System.Reflection.LoaderAllocator

Found 1 roots.
```

The failing tests are:
JIT/Performance/CodeQuality/BenchmarksGame/k-nucleotide/k-nucleotide-9
JIT/Performance/CodeQuality/BenchmarksGame/spectralnorm/spectralnorm-3
JIT/Performance/CodeQuality/BenchmarksGame/mandelbrot/mandelbrot-7
JIT/Performance/CodeQuality/Roslyn/CscBench
JIT/Regression/JitBlue/GitHub_19361/GitHub_19361
JIT/superpmi/superpmicollect/CscBench
profiler/multiple/multiple
Regressions/coreclr/GitHub_45929/test45929
tracing/eventlistener/eventlistener

**Comment by janvorli** [comment]
@jkotas can you please take a look when you have a chance?

**Comment by jkotas** [comment]
I have opened #80450 on this.

**Pull request merged**